system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -175,24 +253,139 @@ export class TopSite extends React.PureComponent {
     </TopSiteLink>);
   }
 }
-TopSite.defaultProps = {link: {}};
+TopSite.defaultProps = {
+  link: {},
+  onDragStart() {}
+};
 
-export const TopSitePlaceholder = () => <TopSiteLink className="placeholder" />;
+export const TopSitePlaceholder = props => <TopSiteLink className="placeholder" isDraggable={false} {...props} />;
 
-export const TopSiteList = props => {
-  const topSites = props.TopSites.rows.slice(0, props.TopSitesCount);
-  const topSitesUI = [];
-  for (let i = 0, l = props.TopSitesCount; i < l; i++) {
-    const link = topSites[i];
-    topSitesUI.push(!link ? <TopSitePlaceholder key={i} /> : <TopSite
-      key={link.guid || link.url}
-      dispatch={props.dispatch}
-      link={link}
-      index={i}
-      intl={props.intl}
-      onEdit={props.onEdit} />);
-  }
-  return (<ul className="top-sites-list">
-    {topSitesUI}
-  </ul>);
-};
+export class TopSiteList extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = this.DEFAULT_STATE = {
+      draggedIndex: null,
+      draggedSite: null,
+      draggedTitle: null,
+      topSitesPreview: null
+    };
+    this.onDragStart = this.onDragStart.bind(this);
+    this.onDragEnd = this.onDragEnd.bind(this);
+    this.onDragEnter = this.onDragEnter.bind(this);
+    this.onDragLeave = this.onDragLeave.bind(this);
+    this.onDrop = this.onDrop.bind(this);
+  }
+  userEvent(event, index) {
+    this.props.dispatch(ac.UserEvent({
+      event,
+      source: TOP_SITES_SOURCE,
+      action_position: index
+    }));
+  }
+  onDragStart(index, link, title) {
+    this.setState({
+      draggedIndex: index,
+      draggedSite: link,
+      draggedTitle: title
+    });
+    this.userEvent("DRAG", index);
+  }
+  onDragEnd() {
+    this.setState(this.DEFAULT_STATE);
+  }
+  onDragEnter(index) {
+    this.setState({topSitesPreview: this._makeTopSitesPreview(index)});
+  }
+  onDragLeave() {
+    this.setState({topSitesPreview: null});
+  }
+  onDrop(index) {
+    this.props.dispatch(ac.SendToMain({
+      type: at.TOP_SITES_DROP,
+      data: {site: {url: this.state.draggedSite.url, label: this.state.draggedTitle}, index}
+    }));
+    // Add a delay to the state reset to avoid ui jank going back to original state
+    // until the pinning happens and the state gets updated and pushed back down.
+    setTimeout(() => this.setState(this.DEFAULT_STATE), 300);
```

Comment:
Typically magic timeouts lead to random intermittent failures in mochitests, but we can at least mock the timer in our tests. For slow machines, the user could see the state reverting then updating when the pin propagates.

Do we need this to reset to default vs optimistically assuming the pin will succeed?

Also, what exactly happens when this gets called and redux state has updated? Does the `setState` end up doing nothing? Setting the state of the site component that happens to reuse it?

@k88hudson any suggestions?

Reply:
Oh, hmm. I guess I can listen for prop updates and clear out the preview once at the point when I get an update with the site properly pinned.

Reply:
I don't completely understand your questions. But basically the preview reorders the sites from the redux store and keeps them in the local state while needed. The sites in the local state takes precedence over the redux store.

Reply:
The thing I was concerned about is related to the lifecycle of the component. Could `this.setState` be called after the props / other state have been changed and something unexpected might happen if we just set it to DEAFULT after 300ms?